### DISCUSSION TITLE
What are the @export and @onready equivalents?

### DISCUSSION THREAD

**Discussion by ullanar** [quaadgras/graphics.gd#43] [Q&A] [answered]
Is it possible to make fields in class structure which will be like gdscript `@export` and `@onready`? Maybe via struct tags? 

```
@export var camera: Camera3D
@onready var camera_mount = $camera_mount
```

**Comment by Splizard** [comment, marked as answer]
Any exported fields in the Go struct will be exported to Godot (just like @export)
https://go.dev/ref/spec#Exported_identifiers

```go
type MyClass struct {
    classdb.Extension[MyClass, Node.Instance]
    
    Camera Camera3D.Instance `gd:"camera"`
}
```

If you add an exported node field to a node, it will be asserted as a child, so in the above example, the Camera will be added as a child node of MyClass.

**Reply by ullanar** [reply]
So this will create child node. But godot `@export` does not do it. It just says what "here is 100% we have object with specified type" and gives u set it in editor on "Node" panel
I assume what there are no such functionality and this is not a very big deal, feature like "super nice to have but ok".

So maybe you can suggest correct way to do proper scene initialization? 

Here for example I want to create `CameraMount` (Node3D) node and then create Camera3D node, save camera to struct and add it to `CameraMount` as child. 


```go
type Player struct {
	classdb.Extension[Player, CharacterBody3D.Instance]
	camera      *Camera3D.Instance
	CameraMount Node3D.Instance `gd:"CameraMount"`
}

func (p *Player) Ready() {
	Input.SetMouseMode(Input.MouseModeCaptured)
	p.camera = Camera3D.New()
	p.CameraMount.AsNode().AddChild(p.camera.AsNode())
	fmt.Printf("camera node name: %v", p.camera.AsNode().GetName())
	p.camera.AsCamera3D().MakeCurrent()
}
```

And here getting error

```
panic: runtime.link/mmm error: use after free

goroutine 17 [running, locked to thread]:
grow.graphics/gd/internal/classdb.Node.GetName({{}, {{}, {0x0?, 0xc000882dc0?}}}, {{0x76163feb1160?, 0xc00043e540?}, 0x76163feb1160?})
        /home/lev/go/pkg/mod/grow.graphics/gd@v0.0.0-20240706100351-36c3a7c07552/internal/classdb/all.go:99484 +0x12d
git.ywa.su/rpp-team/rpp-game-client/pkg/player.(*Player).Ready(0xc0001a4360, {{0x76163efb6e20?, 0xc00043e540?}, 0x76163feb1160?})
        /home/lev/prj/ywa/rpp-game-client/pkg/player/player.go:18 +0x134
grow.graphics/gd.(*instanceImplementation).ready(0xc000012330)
        /home/lev/go/pkg/mod/grow.graphics/gd@v0.0.0-20240706100351-36c3a7c07552/register_class.go:574 +0x54a
grow.graphics/gd.(*instanceImplementation).Notification(0xc000012330, 0xd, 0x16?)
        /home/lev/go/pkg/mod/grow.graphics/gd@v0.0.0-20240706100351-36c3a7c07552/register_class.go:495 +0x37
grow.graphics/gd/gdextension.notification_func(0x0?, 0xd, 0x0)
        /home/lev/go/pkg/mod/grow.graphics/gd@v0.0.0-20240706100351-36c3a7c07552/gdextension/gdextension_interface.go:2496 +0xa3
signal: aborted (core dumped)
```

The reason I want to add `camera` into Player struct is I hate `GetChild`. It is nice but why do u need to find child if u can just take it from struct xD

**Reply by Splizard** [reply]
Sure, try this:
```go
type Player struct {
    classdb.Extension[Player, CharacterBody3D.Instance]

    CameraMount struct {
        Node3D.Instance

        Camera Camera3D.Instance
    } 
}

func (p *Player) Ready() { 
    Input.SetMouseMode(Input.MouseModeCaptured)
    p.CameraMount.Camera.MakeCurrent()
}
```

**Reply by mdevnull** [reply]
> So this will create child node.

Wait, will it? I was pretty sure that it expects the node to be there in the scene tree. If an node of the given name does not exist I did run into issues as well.

**Reply by ullanar** [reply]
> > So this will create child node.
> 
> Wait, will it? I was pretty sure that it expects the node to be there in the scene tree. If an node of the given name does not exist I did run into issues as well.

As I managed to understand: u add field into struct and if it is `gd.Class` and field is public (starting with upper case letter) then child node will be immidiately created on code build. You can even do something with it in editor :)

I truly missing godot `@export` functionality but not that critical.

**Reply by ullanar** [reply]
> Sure, try this:
> 
> ```go
> type Player struct {
>     gd.Class[Player, gd.CharacterBody3D]
> 
>     CameraMount struct {
>         gd.Node3D
> 
>         Camera gd.Camera3D
>     } 
> }
> 
> func (p *Player) Ready() { 
>     gd.Input(p.Temporary).SetMouseMode(gd.InputMouseModeCaptured)
>     p.CameraMount.Camera.MakeCurrent()
> }
> ```

Thank you!

**Comment by ullanar** [comment]
@Splizard 

```go
type Player struct {
    classdb.Extension[Player, CharacterBody3D.Instance]

    CameraMount struct {
        Node3D.Instance

        Camera Camera3D.Instance
    } 
}
```

Thank you for provided suggestion!

I see what it creates 3 separated nodes under Player node: "CameraMount" (node3d), "Node3D" (node3d) and "Camera" (camera3d).

For now it looks a bit like a bug or it is correct behavior? Also Camera is not a child of CameraMount. And I just cant understand what creating that "Node3D" node. If it is CameraMount struct representation then why "CameraMount" even was created

If it is correct behavior - I a bit misunderstand logic here, maybe you will be able to explain?

**Reply by Splizard** [reply]
Ah this is a bug, I'll check on this over the weekend.

**Reply by ullanar** [reply]
One day I really want to help you with development. This is such amazing project for both Godot and Go ecosystems. Yet my lib knowledge is too low I will learn but one day surely will suggest some PRs :)

**Reply by Splizard** [reply]
@Ullanar thank you for your feedback! I've fixed this bug in master, please `go get graphics.gd@master`. 
You should see the following structure:
<img width="277" alt="Screenshot 2024-07-12 at 11 05 23 PM" src="https://github.com/user-attachments/assets/fade4838-7929-489f-b736-8eb8f2523095">
